src/steamship/agents/agents.py

```diff
@@ -151,37 +119,175 @@ def execute_action(self, action: Action, context: AgentContext):
+        # TODO: does this need to be an agent, or can/should we use the tool directly
+        # direct tool use could save on costs, latency, etc.
+        self._personality_agent = ReACTAgent(tools=[personality_tool])
+
+        # Setup our "actions" tool (for doing user prompting)
+        dalle = DalleTool(client=self.client)
+        pix2pix = PixToPixTool(client=self.client)
+        search = SerpTool(client=self.client)
+        prompt_optimizer = PromptOptimizerTool(client=self.client)
+        self._generate_agent = ReACTAgent(
+            client=self.client, tools=[dalle, pix2pix, search, prompt_optimizer]
+        )
+
+    def create_response(self, incoming_message: ChatMessage) -> Optional[List[ChatMessage]]:
+
+        chat_history = ChatMessageHistory(uuid=incoming_message.get_chat_id())
+        chat_history.add_user_message(incoming_message.text)
+
+        # ideally, this Tracing() bit is built elsewhere and passed in, but for now, conform to the
+        # existing `create_response` signature.
+        tracing_ctx = TracingContext.new_context(AgentContext(), Tracing())
+        chat_ctx = ChatContext.new_context(tracing_ctx, chat_file=chat_history)
+
+        # todo: allow read-only access to history here?
+        # does the agent need full conversation history for tool selection? Is that potentially problematic?
+        output_blocks = self._generate_agent.run(agent_input=[incoming_message], context=chat_ctx)
+
+        responses = []
+        for block in output_blocks:
+            if block.is_text():
+                # do not use the chat history here, as context is not needed for personality translation
+                translated_blocks = self._personality_agent.run(
+                    agent_input=[block], context=tracing_ctx
+                )
+                if len(translated_blocks) == 1:
+                    # update chat history here, to allow for ai-user convo saving
+                    chat_history.add_ai_message(translated_blocks[0].text)
+                    responses.append(
+                        ChatMessage.from_block(
+                            translated_blocks[0],
+                            chat_id=incoming_message.get_chat_id(),
+                            message_id=incoming_message.get_message_id(),
+                        )
+                    )
+            else:
+                # do we want to update chat history with multi-media representation?  i _think_ so?
+                chat_history.add_generated_block(block)
+                # todo: decide if you want personality driven captions, etc.
+                responses.append(
+                    ChatMessage.from_block(
+                        block,
+                        chat_id=incoming_message.get_chat_id(),
+                        message_id=incoming_message.get_message_id(),
+                    )
+                )
+
+        return responses
+
+
+class DalleTool(Tool):
```

Comment:
@douglas-reid before you implement sketched of these these, take a look at my open PR in case you want to borrow code!